src/Microsoft.AspNet.Mvc.Localization/HtmlLocalizer.cs

```diff
@@ -0,0 +1,415 @@
+        protected string EncodeArguments([NotNull] string resourceString, [NotNull] object[] arguments)
+        {
+            var position = 0;
+            var length = resourceString.Length;
+            var ch = '\x0';
+            StringBuilder builder = null;
+            var resourceStringBuilder = new StringBuilder();
+
+            while (true)
+            {
+                builder = new StringBuilder();
+
+                var isToken = false;
+                var p = position;
+                var i = position;
+
+                while (position < length)
+                {
+                    ch = resourceString[position];
+
+                    position++;
+                    if (ch == '}')
+                    {
+                        if (position < length && resourceString[position] == '}') // Treat as escape character for }}
+                        {
+                            position++;
+                            if (position < length && resourceString[position] != '{')
+                            {
+                                isToken = false;
+                                builder.Append(ch);
+                                resourceStringBuilder.Append(EncodeArgument(builder.ToString()));
+                                builder = new StringBuilder();
+                            }
+                        }
+                        else
+                        {
+                            FormatError();
+                        }
+                    }
+
+                    if (ch == '{')
+                    {
+                        if (builder == null)
+                        {
+                            builder = new StringBuilder();
+                        }
+                        if (position < length && resourceString[position] == '{') // Treat as escape character for {{
+                        {
+                            isToken = true;
+                            position++;
+                        }
+                        else
+                        {
+                            position--;
+                            break;
+                        }
+                    }
+
+                    if (isToken)
+                    {
+                        builder.Append(ch);
+                    }
+                    else
+                    {
+                        resourceStringBuilder.Append(ch);
+                    }
+                }
+
+                if (position == length)
+                {
+                    if (builder.Length > 0)
+                    {
+                        resourceStringBuilder.Append(EncodeArgument(builder.ToString()));
+                    }
+                    break;
+                }
+                position++;
+                if (position == length || (ch = resourceString[position]) < '0' || ch > '9')
+                {
+                    FormatError();
+                }
+
+                var index = 0;
```

Comment:
Why parse the bit between the curly braces at all? I expected to see code that handles `'{'`, `'}'` and escaped versions of those characters but nothing else e.g.
``` c#
var isToken = false;
while (position < length)
{
    var currentChar = resourceString[position];
    position++;
    if (currentChar == '{')
    {
        // Handle escaped curly; copy as appropriate
        // Check isToken, error if already true, et cetera
        // In most cases, set isToken
    }
    else if (currentChar == '}')
    {
        // Handle escaped curly; copy as appropriate
        // In most cases, reset isToken
        if (tokenBuffer != null && tokenBuffer.Length > 0)
        {
             var formattedToken = string.Format(tokenBuffer, params);
             outputBuffer.Append(Encode(formattedToken);
             tokenBuffer = null;
        }
    }
    else if (isToken)
    {
        // Copy into token buffer
    }
    else
    {
        // Copy into output buffer.
    }
}

if (tokenBuffer != null && tokenBuffer.Length > 0)
{
     var formattedToken = string.Format(tokenBuffer, params);
     outputBuffer.Append(Encode(formattedToken);
     tokenBuffer = null;
}
```